providers/implementations/keymgmt/ecx_kmgmt.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2020 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <assert.h>
+#include <openssl/core_numbers.h>
+#include <openssl/core_names.h>
+#include <openssl/params.h>
+#include "internal/param_build.h"
+#include "crypto/ecx.h"
+#include "prov/implementations.h"
+#include "prov/providercommon.h"
+
+static OSSL_OP_keymgmt_importkey_fn x25519_import_key;
+static OSSL_OP_keymgmt_importkey_fn x448_import_key;
+static OSSL_OP_keymgmt_exportkey_fn x25519_export_key;
+static OSSL_OP_keymgmt_exportkey_fn x448_export_key;
+static OSSL_OP_keymgmt_get_key_params_fn x25519_get_key_params;
+static OSSL_OP_keymgmt_get_key_params_fn x448_get_key_params;
+
+static void *ecx_importkey(void *provctx, const OSSL_PARAM params[],
+                           size_t keylen)
+{
+    ECX_KEY *key = NULL;
+    size_t privkeylen = 0, pubkeylen;
+    const OSSL_PARAM *param_priv_key, *param_pub_key;
+    unsigned char *pubkey;
+
+    param_priv_key =
+        OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_ECX_PRIV_KEY);
+    param_pub_key =
+        OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_ECX_PUB_KEY);
+
+    /*
+     * If a private key is present then a public key must also be present.
+     * Alternatively we've just got a public key.
+     */
+    if (param_pub_key == NULL)
+        return NULL;
+
+    key = ecx_key_new(keylen);
+    if (key == NULL)
+        return NULL;
+
+    pubkey = key->pubkey;
+    if ((param_priv_key != NULL
+             && !OSSL_PARAM_get_octet_string(param_priv_key,
+                                            (void **)&key->privkey, keylen,
+                                             &privkeylen))
+            || !OSSL_PARAM_get_octet_string(param_pub_key,
```

Comment:
Hmm this might be more readable if there are 2 if's here..
One that just handles the  private bit and does the privlen check
And one that does the public bit and public length check.

Reply:
Done